### ISSUE TITLE
[type hierarchy] Remove confusing methods from Traversable that treat it as 0..1 container

### ISSUE THREAD

**Issue by danieldietrich** [vavr-io/vavr#2023]
There will still be changes to the type hierarchy in the next major release Vavr 1.0.0. For example #1295

We should remove all methods from `Traversable` that treat a collection as container that has 0..1 elements.

For example @Mousaka stated:

> when/how can I use `Map`'s `getOrElseThrow(Supplier)` ? What does it mean to get something from a Map without giving a key?

He is absolutely right - this is an uncommon collection behaviour. We are able to achieve the goal of removing the mentioned methods from Traversable by performing the following changes:

1. Traversable does not extend Value anymore
2. We factor out commonalities between Value and Traversable into a non-public interface that is extended by both

**Comment by danieldietrich** [comment]
Uncomment the following code block in `io.vavr.PartialFunction` after this issue is solved:

```java
    /* TODO: currently the name is ambiguous. uncomment after #2023 is finished!
    default PartialFunction<T, R> orElse(PartialFunction<? super T, ? extends R> fallback) {
        return of(t -> applyOrElse(t, fallback), t -> isDefinedAt(t) || fallback.isDefinedAt(t));
    }
    */
```

**Comment by danieldietrich** [comment]
Already done in master (= vavr 1.0 branch) by removing Value interface.
